File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    char[][] grid;
    int rows;
    int cols;
    boolean[][] visited;

    public int numIslands(char[][] grid) {
        this.grid = grid;
        rows = grid.length;
        cols = grid[0].length;
        visited = new boolean[rows][cols];

        int count = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    count++;
                    dfs(i, j);
                }
            }
        }

        return count;
    }
    
    private void dfs(int i, int j) {
        if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] == '0' || visited[i][j]) return;
        visited[i][j] = true;
        dfs(i + 1, j);
        dfs(i - 1, j);
        dfs(i, j + 1);
        dfs(i, j - 1);
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-1.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n1, n2 = len(s), len(t)
        dp = {}

        def dfs(i1, i2):
            if not (i1, i2) in dp:
                if i2 == n2:
                    dp[(i1, i2)] = 1
                elif i1 == n1:
                    dp[(i1, i2)] = 0
                elif s[i1] == t[i2]:
                    dp[(i1, i2)] = dfs(i1 + 1, i2 + 1) + dfs(i1 + 1, i2)
                else:
                    dp[(i1, i2)] = dfs(i1 + 1, i2)
            return dp[(i1, i2)]
        
        return dfs(0, 0)